### PR TITLE
Fix: Enable Text Align UI to be controlled correctly with theme.json

### PR DESCRIPTION
**Pull request by t-hamano**
Follow-up #59531

## What?

This PR allows `textAlign` block support to be controlled via theme.json.

## Why?

We added the new `textAlign` block support in #59531. However, since enabling this support depended _only_ on block support, it turned out that it could not be disabled via theme.json. In other words, we cannot disable this block support even if you write code like the following:

block.json

```json
{
	"apiVersion": 3,
	"name": "core/test",
	"title": "Test",
	"supports": {
		"typography": {
			"textAlign": true
		}
	}
}
```

theme.json

```json
{
	"version": 3,
	"settings": {
		"typography": {
			"textAlign": false
		}
	}
}
```

## How?

Checks whether theme.json and block.json are supported, and displays the block toolbar UI if both support it. This logic is implemented as a `useIsTextAlignDisabled` hook and is similar to the logic in [useIsFontSizeDisabled](https://github.com/WordPress/gutenberg/blob/ea946fdf9aa39d1504c1027c2ba8b8435e87fc10/packages/block-editor/src/hooks/font-size.js#L141-L148), [useIsLineHeightDisabled](https://github.com/WordPress/gutenberg/blob/ea946fdf9aa39d1504c1027c2ba8b8435e87fc10/packages/block-editor/src/hooks/line-height.js#L56-L62), etc.

## Testing Instructions

Temporarily add this support to the Media & Text block and the Group block:

<details><summary>Diff</summary>

```diff
diff --git a/packages/block-library/src/group/block.json b/packages/block-library/src/group/block.json
index db7d09c55d..444cfd41ec 100644
--- a/packages/block-library/src/group/block.json
+++ b/packages/block-library/src/group/block.json
@@ -73,6 +73,7 @@
                        "sticky": true
                },
                "typography": {
+                       "textAlign": true,
                        "fontSize": true,
                        "lineHeight": true,
                        "__experimentalFontFamily": true,
diff --git a/packages/block-library/src/media-text/block.json b/packages/block-library/src/media-text/block.json
index 0f3519acb5..782da2e1eb 100644
--- a/packages/block-library/src/media-text/block.json
+++ b/packages/block-library/src/media-text/block.json
@@ -117,6 +117,7 @@
                        "padding": true
                },
                "typography": {
+                       "textAlign": true,
                        "fontSize": true,
                        "lineHeight": true,
                        "__experimentalFontFamily": true,
```

</details> 

### Default (Classic Theme)

- Activate Twenty Twenty-One.
- The UI should be visible in both blocks.

### Default (Theme with theme.json)

- Activate Emptytheme.
- Define an empty theme.json.
  ```json
  {
  	"$schema": "https://schemas.wp.org/trunk/theme.json",
  	"version": 3,
  	"settings": {
  		"layout": {
  			"contentSize": "840px"
  		}
  	}
  }
  ```
- The UI should be visible in both blocks.
 
### Disable at root level of theme.json

- Activate Emptytheme.
- Define theme.json as below.
  ```json
  {
  	"$schema": "https://schemas.wp.org/trunk/theme.json",
  	"version": 3,
  	"settings": {
  		"layout": {
  			"contentSize": "840px"
  		},
  		"typography": {
  			"textAlign": false
  		}
  	}
  }
  ```

- The UI should NOT be visible in both blocks.

### Disable at block level of theme.json

- Activate Emptytheme.
- Define theme.json as below.
  ```json
  {
  	"$schema": "https://schemas.wp.org/trunk/theme.json",
  	"version": 3,
  	"settings": {
  		"layout": {
  			"contentSize": "840px"
  		},
  		"blocks": {
  			"core/media-text": {
  				"typography": {
  					"textAlign": false
  				}
  			}
  		}
  	}
  }
  ```

- Only the Media and Text block should not display this UI.

### CODE REVIEW

**Comment by t-hamano** [comment]
P.S. Both #61182 and #59531 need to be backported to the core, so give them a label.

File: packages/block-editor/src/hooks/text-align.js [hunk 1]
```diff
@@ -70,11 +71,17 @@ function BlockEditTextAlignmentToolbarControlsPure( {
 	name: blockName,
 	setAttributes,
 } ) {
+	const isDisabled = useIsTextAlignDisabled( blockName );
```
Comment (aaronrobertshaw) [inline packages/block-editor/src/hooks/text-align.js]: I think we might be able to use `useBlockSettings` here instead of creating a new hook.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for catching this one and working on the fix @t-hamano 👍 

I think we might be re-inventing the wheel a bit with `useIsTextAlignDisabled`. 

When all the block support panels were refactored so the post and site editors share the same component, the approach to checking support and whether the UI was disabled, was also merged. 

My understanding is that the theme.json settings are retrieved and then the block specific supports are merged over those. 

Take a look at:
- [useBlockSettings](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/utils.js?plain=1#L219)
- [useSettingsForBlockElement](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/global-styles/hooks.js?plain=1#L218)
- [getSupportedStyles](https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/src/store/private-selectors.js?plain=1#L99)

Also, I think we need to update the title of this PR to reflect what it's actually about i.e. disabling UI controls rather than the actual inheritance of values etc.

**Comment by t-hamano** [comment]
@aaronrobertshaw Thanks for the review!

> I think we might be able to use `useBlockSettings` here instead of creating a new hook.

Admittedly, using this hook is simpler 👍

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for applying the suggested tweaks @t-hamano. Looks good 👍 

This still tests well for me:

✅ Classic Theme: UI is visible for both blocks
✅ Block Theme (default): UI is visible for both blocks
✅ Block Theme (disabled): UI is NOT visible for both blocks
✅ Block Theme (disabled by block type): Disabling for individual blocks works as expected

Nit: While reading through the code for the text align hook to refresh my memory, it stuck out that the `getValidTextAlignments` function is a bit long-winded and could be simplified.

<details>
<summary>Maybe we could tweak that with something like this?</summary>

```diff
diff --git a/packages/block-editor/src/hooks/text-align.js b/packages/block-editor/src/hooks/text-align.js
index d1edadfad5..1a0cfe2ecc 100644
--- a/packages/block-editor/src/hooks/text-align.js
+++ b/packages/block-editor/src/hooks/text-align.js
@@ -43,6 +43,7 @@ const TEXT_ALIGNMENT_OPTIONS = [
 ];
 
 const VALID_TEXT_ALIGNMENTS = [ 'left', 'center', 'right' ];
+const NO_TEXT_ALIGNMENTS = [];
 
 /**
  * Returns the valid text alignments.
@@ -54,19 +55,12 @@ const VALID_TEXT_ALIGNMENTS = [ 'left', 'center', 'right' ];
  * @return {string[]} Valid text alignments.
  */
 export function getValidTextAlignments( blockTextAlign ) {
-	let validTextAlignments;
 	if ( Array.isArray( blockTextAlign ) ) {
-		validTextAlignments = VALID_TEXT_ALIGNMENTS.filter( ( textAlign ) =>
+		return VALID_TEXT_ALIGNMENTS.filter( ( textAlign ) =>
 			blockTextAlign.includes( textAlign )
 		);
-	} else if ( blockTextAlign === true ) {
-		// `true` includes all alignments...
-		validTextAlignments = [ ...VALID_TEXT_ALIGNMENTS ];
-	} else {
-		validTextAlignments = [];
 	}
-
-	return validTextAlignments;
+	return blockTextAlign === true ? VALID_TEXT_ALIGNMENTS : NO_TEXT_ALIGNMENTS;
 }
 
 function BlockEditTextAlignmentToolbarControlsPure( {

```

</details>

**Comment by t-hamano** [comment]
@aaronrobertshaw Thanks for the review!

> Maybe we could tweak that with something like this?

This is certainly easier to understand. [The unit test for the function](https://github.com/WordPress/gutenberg/blob/608b8900dc7a49cf8585d3027798ee374acad87d/packages/block-editor/src/hooks/test/text-align.js#L31-L62) passed locally, so the logic shouldn't have changed at all.

**Comment by t-hamano** [comment]
Ready to backport to the core.

- Ticket: https://core.trac.wordpress.org/ticket/61256
- PR: https://github.com/WordPress/wordpress-develop/pull/6590

**Pull request merged**